Step 1. Data preperation
==============================================

**Author**: Philipp Muench (pmu15@helmholtz-hzi.de)

**Date**: `r Sys.Date()`

----------------------------------------

<style type="text/css">
h1 {
	line-height: 120%;
}
</style>

# prepeare Hidden Markov Models

First we need to specify [HMM profiles](https://en.wikipedia.org/wiki/HMMER) for the proteins we want to identify. These were searched using the [Pfam website](https://pfam.xfam.org/search#tabview=tab2) for involved genes from [Ry. Young, 2014](https://www.ncbi.nlm.nih.gov/pmc/articles/pmid/24585055/)

- R encodes the endolysin
- Rz and Rz1 are nested genes that encode outer membrane proteins that
are auxiliary lysis factors.

```{r, echo=F, results='asis'}
library(knitr)
table1 <- read.table("table1.csv", header=T, sep=";")
kable(table1, caption = "gene families for prophage lysis-genes")
```

List of colicins are taken from https://de.wikipedia.org/wiki/Colicine. UniRef terms serached via [UniProt](https://www.uniprot.org/uniref/) with the search term _uniref50 "colicin B"_


```{r, echo=F, results='asis'}
library(knitr)
table3 <- read.table("table3.csv", header=T, sep=";")
kable(table3, caption = "Uniref for colicin genes")
```

The corresponding .fasta files are downloaded using the search term _cluster:UniRef50_P09883 AND identity:1.0_ where the UniRef ID is the corresponding ID from the table and saved to `hmm/colicin/faa`. These files were aligned using [Clustal Omega](https://www.ebi.ac.uk/Tools/msa/clustalo/) using standard parameters and the alignment was saved to the `hmm/colicin/msa` folder. [Hmmer](http://hmmer.org/) was used to generate a HMM profile from the MSA and was saved to the `hmm/colicin/hmm` folder e.g using the command `hmmbuild hmm/UniRef50_UPI00022A9A6A.hmm msa/UniRef50_UPI00022A9A6A.clustal_num`. 

The full HMM models for the Pfam collection (v. 32) can be downloaded via EBI. This model consists of a collection of individual HMM profiles that can be extracted from this `.hmm` file. These individual HMM files have been placed to the `hmm/` folder (e.g. `hmm/PF04971.hmm`)

```{bash, eval=F}
wget ftp://ftp.ebi.ac.uk/pub/databases/Pfam/releases/Pfam32.0/Pfam-A.hmm.gz
gunzip Pfam-A.hmm.gz
```

after extracting the individual HMM files we join them together
```{bash, eval=F}
cat hmm/*.hmm > hmm/lysis.hmm
cat hmm/colicin/hmm/*.hmm > hmm/colicin.hmm
```


# prepare Salmonella genomes

Now download all genomes form the [NCBI genome browser](https://www.ncbi.nlm.nih.gov/genome/browse/) using the search term _"Salmonella enterica"_ and filter for _"Prokaryotes"_ resulting in 9754 entries. Downloadable list are saved as `ncbi_list.csv`. Then we remove the first character from this file using following command.

```{bash, eval=F}
sed '1s/^.//' ncbi_list.csv > ncbi_list_with_header.csv
```

and read it to R and save the ftp location to the file `ftp_links.txt`

```{r}
dat <- read.table("ncbi_list_with_header.csv", header=T, sep=",")
write.table(dat$RefSeq.FTP, file="ftp_links.txt", quote=F, col.names=F, row.names=F)
```

now we can use `wget` to download all genomic fasta files

```{bash, eval=F}
mkdir -p genomes/faa
wget -i ftp_links.txt -r -np -nd -A "*_protein.faa.gz" -P genomes/faa
gunzip genomes/faa/*.gz
```

# detection of prophages 

We use PhiSpy ([paper](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3439882/)) to detect prophages in genomes. For this we have to download genebank entries for these genomes

```{bash, eval=F}
mkdir -p genomes/gb
wget -i ftp_links.txt -r -np -nd -A "*_genomic.gbff.gz" -P genomes/gb
gunzip genomes/gb/*.gz
```

```{bash, eval=F}
mkdir -p tools
cd tools
git clone https://github.com/linsalrob/PhiSpy.git
cd PhiSpy/ 
make
```

We need to reformat the genomes to the seed format, for that we use the genebank `.gbff` files downloaded before. For this step you should have [GNU parallel](https://www.gnu.org/software/parallel/) installed `sudo apt-get install parallel`

```{bash, eval=F}
mkdir -p genomes/seed
parallel 'python tools/PhiSpy/genbank_to_seed.py {} {.}.seed' ::: genomes/gb/*.gbff
mv genomes/gb/*.seed genomes/seed
```

nor we run PhiSpy on all seed files
```{bash, eval=F}
parallel 'python tools/PhiSpy/PhiSpy.py -i {} -o {.}.phispy_out' ::: genomes/seed/*.seed
mkdir -p genomes/phispy
mv genomes/seed/*.phispy_out genomes/phispy
```

## Sessioninfo

```{r}
sessionInfo()
```

